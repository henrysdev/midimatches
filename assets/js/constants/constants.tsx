import * as Tone from "tone";

// MIDI codes
export const NOTE_ON = 0x9;
export const NOTE_OFF = 0x8;

// Possible game state views
export enum GAME_VIEW {
  PREGAME_LOBBY,
  GAME_START,
  ROUND_START,
  RECORDING,
  PLAYBACK_VOTING,
  ROUND_END,
  GAME_END,
}

export enum PRACTICE_GAME_VIEW {
  SAMPLE_SELECTION,
  RECORDING,
  PLAYBACK,
}

export const S3_BUCKET_URL = "https://progressions-game.s3.amazonaws.com";

export const SERVERLIST_UPDATE_EVENT = "serverlist_update";

export const GAME_UPDATE_EVENT = "game_update";
export const START_GAME_EVENT = "start_game";
export const RESET_ROOM_EVENT = "reset_room";
export const LOBBY_UPDATE_EVENT = "lobby_update";
export const NEW_CHAT_MESSAGE_EVENT = "new_chat_message";

export const SUBMIT_REGISTERED_JOIN = "user_enter_room";
export const SUBMIT_JOIN = "player_join";
export const SUBMIT_LEAVE_ROOM = "player_leave_room";
export const SUBMIT_READY_UP_EVENT = "player_ready_up";
export const SUBMIT_RECORDING_EVENT = "player_recording";
export const SUBMIT_VOTE_EVENT = "player_vote";
export const SUBMIT_CHAT_MESSAGE = "player_chat_message";

export const MAX_CHAT_HISTORY_LENGTH = 200;
export const MAX_CHARS_PER_CHAT_MESSAGE = 80;

export const MIN_PLAYER_ALIAS_LENGTH = 3;
export const MAX_PLAYER_ALIAS_LENGTH = 10;

export const DEFAULT_SAMPLE_COLORS = [
  "rgba(255, 127, 0, 0.15)",
  "rgba(255, 213, 0, 0.15)",
  "rgba(255, 212, 0, 0.15)",
  "rgba(106, 255, 0, 0.15)",
  "rgba(0, 234, 255, 0.15)",
  "rgba(0, 149, 255, 0.15)",
  "rgba(0, 64, 255, 0.15)",
  "rgba(170, 0, 255, 0.15)",
  "rgba(255, 0, 170, 0.15)",
  "rgba(143, 35, 35, 0.15)",
  "rgba(143, 106, 35, 0.15)",
  "rgba(79, 143, 35, 0.15)",
  "rgba(35, 98, 143, 0.15)",
  "rgba(107, 35, 143, 0.15)",
];

export const DEFAULT_SAMPLE_PLAY_BUFFER_LENGTH = 2;
export const DEFAULT_SAMPLE_LENGTH = 10.6667;
export const DEFAULT_NUM_WARMUP_LOOPS = 1;
export const DEFAULT_NUM_RECORDED_LOOPS = 1;
export const DEFAULT_RECORDING_LENGTH =
  DEFAULT_SAMPLE_LENGTH * DEFAULT_NUM_RECORDED_LOOPS;
export const DEFAULT_WARMUP_LENGTH =
  DEFAULT_SAMPLE_LENGTH * DEFAULT_NUM_WARMUP_LOOPS;

export const DEFAULT_SAMPLE_VOLUME = -4;

export const DEFAULT_SYNTH_CONFIG = {
  oscillator: {
    type: "amtriangle",
    harmonicity: 0.5,
    modulationType: "sine",
  },
  envelope: {
    attackCurve: "exponential",
    attack: 0.03,
    decay: 0.4,
    sustain: 0.2,
    release: 0.5,
  },
  volume: -2,
  portamento: 0.05,
} as Tone.SynthOptions;

export const DEFAULT_FM_SYNTH_CONFIG = {
  volume: -3,
  detune: 0,
  portamento: 0,
  harmonicity: 1,
  oscillator: {
    partialCount: 0,
    partials: [],
    phase: 0,
    type: "sine",
  },
  envelope: {
    attack: 0.001,
    attackCurve: "linear",
    decay: 0.3,
    decayCurve: "exponential",
    release: 0.5,
    releaseCurve: "exponential",
    sustain: 1,
  },
  modulation: {
    partialCount: 0,
    partials: [],
    phase: 7,
    type: "sawtooth",
  },
  modulationEnvelope: {
    attack: 0.03,
    attackCurve: "linear",
    decay: 0.01,
    decayCurve: "exponential",
    release: 0.5,
    releaseCurve: "exponential",
    sustain: 1,
  },
  modulationIndex: 12.22,
} as any;

export const DEFAULT_MANUAL_NOTE_VELOCITY = 65;

// room creation parameters
export const MIN_ROOM_SIZE = 3;
export const MAX_ROOM_SIZE = 6;
export const DEFAULT_ROOM_SIZE = 4;

export const MIN_NUM_ROUNDS = 1;
export const MAX_NUM_ROUNDS = 10;
export const DEFAULT_NUM_ROUNDS = 3;

export const MIN_ROOM_NAME_LENGTH = 3;
export const MAX_ROOM_NAME_LENGTH = 20;

export const MIN_SOUND_VOLUME = -25.0;
export const MAX_SOUND_VOLUME = 2.5;

export const SOUND_VOLUME_COOKIE = "midimatches_soundvolume";
