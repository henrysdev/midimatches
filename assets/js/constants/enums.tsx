export const enum GAME_VIEW {
  START,
  INPUT_TEST,
  SAMPLE_RECORDING,
  SAMPLE_PLAYBACK,
  VOTING,
  RESULTS
}