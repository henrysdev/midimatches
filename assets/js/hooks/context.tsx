import { useContext } from "react";

import {
  GameContext,
  ToneAudioContext,
  PlayerContext,
  CurrentUserContext,
  SocketContext,
  ChatContext,
} from "../contexts";

import {
  GameContextType,
  ToneAudioContextType,
  PlayerContextType,
  CurrentUserContextType,
  SocketContextType,
  ChatContextType,
} from "../types";

export function useGameContext() {
  return useContext(GameContext) as GameContextType;
}

export function useToneAudioContext() {
  return useContext(ToneAudioContext) as ToneAudioContextType;
}

export function usePlayerContext() {
  return useContext(PlayerContext) as PlayerContextType;
}

export function useCurrentUserContext() {
  return useContext(CurrentUserContext) as CurrentUserContextType;
}

export function useSocketContext() {
  return useContext(SocketContext) as SocketContextType;
}

export function useChatContext() {
  return useContext(ChatContext) as ChatContextType;
}
