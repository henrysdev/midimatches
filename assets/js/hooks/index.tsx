export { useSamplePlayer } from "./useSamplePlayer";
export { useGameServerState } from "./useGameServerState";
export { useWebMidi } from "./useWebMidi";
export * from "./context";
