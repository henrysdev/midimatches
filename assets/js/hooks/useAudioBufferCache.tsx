import { useState, useEffect, useMemo } from "react";
import { ToneAudioBuffer } from "tone";

type AudioBufferCacheTuple = [AudioBufferCache, (urls: string[]) => void];

type AudioBufferCache = Map<string, ToneAudioBuffer>;

export function useAudioBufferCache(): AudioBufferCacheTuple {
  const [bufferCache, setBufferCache] = useState<AudioBufferCache>(
    new Map<string, ToneAudioBuffer>()
  );

  const populateBufferCache = async (urls: string[]) => {
    await Promise.all(
      urls.map(async (url) => {
        if (!bufferCache.has(url)) {
          const newBuf = new ToneAudioBuffer();
          await newBuf.load(url);
          setBufferCache((oldCache) => {
            return oldCache.set(url, newBuf);
          });
        }
      })
    );
  };

  return [bufferCache, populateBufferCache];
}
