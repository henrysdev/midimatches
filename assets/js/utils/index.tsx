export * from "./transforms";
export * from "./time";
