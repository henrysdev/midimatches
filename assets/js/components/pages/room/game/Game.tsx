import { Channel } from "phoenix";
import React, { useEffect, useState, useMemo } from "react";

import * as Tone from "tone";
import { GAME_VIEW, DEFAULT_SYNTH_CONFIG } from "../../../../constants";
import { GameContext, ToneAudioContext } from "../../../../contexts";
import {
  GameEndView,
  GameStartView,
  PlaybackVotingView,
  RecordingView,
  RoundEndView,
  RoundStartView,
} from "./views";
import {
  useGameServerState,
  useSamplePlayer,
  usePlayerContext,
} from "../../../../hooks";
import { GameLayout, InGameFrame } from ".";
import { Input } from "webmidi";
import { GameContextType } from "../../../../types";

interface GameProps {
  gameChannel: Channel;
  initGameState: GameContextType;
}

const Game: React.FC<GameProps> = ({ gameChannel, initGameState }) => {
  const [currentView, gameContext, pushMessage] = useGameServerState(
    gameChannel,
    initGameState
  );
  const [midiInputs, setMidiInputs] = useState<Array<Input>>([]);
  const [loadSample, playSample, stopSample] = useSamplePlayer(Tone);
  const { player: currPlayer } = usePlayerContext();
  const [synth, setSynth] = useState<any>();

  const currSampleBeat = useMemo(() => {
    return gameContext.sampleBeats[gameContext.roundNum - 1];
  }, [gameContext.roundNum]);

  const resetTone = () => {
    stopSample();
    Tone.Transport.cancel(0);
    Tone.Transport.stop();
  };

  useEffect(() => {
    Tone.context.lookAhead = 0;
    const newSynth = new Tone.PolySynth(DEFAULT_SYNTH_CONFIG).toDestination();
    // const newSynth = new Tone.Sampler({
    //   urls: {
    //     C4: "funk_daddy_c4.mp3",
    //     C5: "funk_daddy_c5.mp3",
    //   },
    //   baseUrl: "https://progressions-game.s3.amazonaws.com/synths/funk_daddy/",
    // }).toDestination();
    setSynth(newSynth);
  }, []);

  useEffect(() => {
    switch (currentView) {
      case GAME_VIEW.GAME_START:
        break;
      case GAME_VIEW.ROUND_START:
        loadSample(currSampleBeat);
        break;
      case GAME_VIEW.RECORDING:
        loadSample(currSampleBeat);
        break;
      case GAME_VIEW.PLAYBACK_VOTING:
        loadSample(currSampleBeat);
        resetTone();
        break;
      case GAME_VIEW.ROUND_END:
        resetTone();
        break;
      case GAME_VIEW.GAME_END:
        resetTone();
        break;
    }
  }, [currentView]);

  return (
    <GameContext.Provider value={gameContext} key={currentView}>
      <ToneAudioContext.Provider value={{ Tone, midiInputs, synth }}>
        <InGameFrame>
          {(() => {
            switch (currentView) {
              case GAME_VIEW.GAME_START:
                return (
                  <GameStartView
                    pushMessageToChannel={pushMessage}
                    setMidiInputs={setMidiInputs}
                  />
                );

              case GAME_VIEW.ROUND_START:
                return (
                  <RoundStartView
                    pushMessageToChannel={pushMessage}
                    roundNum={gameContext.roundNum}
                  />
                );

              case GAME_VIEW.RECORDING:
                return (
                  <RecordingView
                    isContestant={
                      !!gameContext.players
                        ? gameContext.players
                            .map(({ playerId }) => playerId)
                            .includes(currPlayer.playerId)
                        : false
                    }
                    pushMessageToChannel={pushMessage}
                    playSample={playSample}
                    stopSample={stopSample}
                  />
                );

              case GAME_VIEW.PLAYBACK_VOTING:
                return (
                  <PlaybackVotingView
                    pushMessageToChannel={pushMessage}
                    playSample={playSample}
                    stopSample={stopSample}
                  />
                );

              case GAME_VIEW.ROUND_END:
                return <RoundEndView />;

              case GAME_VIEW.GAME_END:
                return <GameEndView />;
            }
          })()}
        </InGameFrame>
      </ToneAudioContext.Provider>
    </GameContext.Provider>
  );
};
export { Game };
