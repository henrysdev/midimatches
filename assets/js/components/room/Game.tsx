import React from "react";
import { useState } from "react";
import { GenericView, RecordingView } from './views/index';
import { Socket } from "phoenix"
import { GAME_VIEW } from '../../constants/index';

interface GameProps {}

const Game: React.FC<GameProps> = () => {
  const [currentView, setCurrentView] = useState(GAME_VIEW.SAMPLE_RECORDING);

  const changeView = (nextView: GAME_VIEW) => {
    setCurrentView(nextView);
  }

  // SOCKET STUFF
  let socket = new Socket("/socket", {params: {token: window.userToken}});
  socket.connect()
  let path              = window.location.pathname.split('/')
  let room_id           = path[path.length -1]
  let channel           = socket.channel(`room:${room_id}`);
  let roomStartTime     = 0;
  channel.join()
  .receive("ok", resp => { console.log("Joined successfully", resp)})
  .receive("error", resp => { console.log("Unable to join", resp)});

  channel.on("init_room_client", ({start_time_utc}) => {
    roomStartTime = start_time_utc;
  });

  channel.on("broadcast_updated_musician_loop", payload => {
    console.log('RECV broadcast_updated_musician_loop', payload);
  });

  channel.on("game_view", payload => {
    console.log("game view: ", payload);
    setCurrentView(payload);
  });

  // channel.push("update_musician_loop", {
  //   loop: JSON.stringify({})
  // });

  switch(currentView) {
    case GAME_VIEW.START:
      console.log("START");
      return (
        <GenericView currentView={currentView} nextViewCallback={() => changeView(GAME_VIEW.INPUT_TEST)} />
      );
    case GAME_VIEW.INPUT_TEST:
      console.log("INPUT_TEST");
      return (
        <GenericView currentView={currentView} nextViewCallback={() => changeView(GAME_VIEW.SAMPLE_RECORDING)} />
      );
    case GAME_VIEW.SAMPLE_RECORDING:
      console.log("SAMPLE_RECORDING");
      return (
        <RecordingView channel={channel} nextViewCallback={() => changeView(GAME_VIEW.SAMPLE_PLAYBACK)} /> 
      );
    case GAME_VIEW.SAMPLE_PLAYBACK:
      console.log("SAMPLE_PLAYBACK");
      return (
        <GenericView currentView={currentView} nextViewCallback={() => changeView(GAME_VIEW.VOTING)} />
      );
    case GAME_VIEW.VOTING:
      console.log("VOTING");
      return (
        <GenericView currentView={currentView} nextViewCallback={() => changeView(GAME_VIEW.RESULTS)} />
      );
    case GAME_VIEW.RESULTS:
      console.log("RESULTS");
      return (
        <GenericView currentView={currentView} nextViewCallback={() => changeView(GAME_VIEW.START)} />
      );
    default:
      console.log('unrecognized event type encountered: ', currentView);
  }
  return (
    <>
    </>
  );
};
export { Game };