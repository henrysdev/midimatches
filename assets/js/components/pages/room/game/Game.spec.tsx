import React from "react";
import { Channel, Socket } from "phoenix";
import { Game } from "./Game";
import { render, fireEvent, screen } from "@testing-library/react";
import { GAME_VIEW } from "../../../../constants";
import { mocked } from "ts-jest/utils";
import { useGameServerState, useToneAudioContext } from "../../../../hooks";

const mockedSocket = new Socket("fake_endpoint");
const mockedChannel = new Channel("fake_topic", {}, mockedSocket);
mockedChannel.join();

const gameRulesMock = {
  gameSizeNumPlayers: 4,
  soloTimeLimit: 30,
  timestepSize: 500,
  quantizationThreshold: 0.6,
};

const pushMessageMock = () => {};

const tonePlayerMock = {
  toDestination: () => tonePlayerMock,
  volume: {
    value: 0,
  },
};
const toneMock = {
  Player: () => tonePlayerMock,
  now: () => 10,
};
const mockGameStartView = GAME_VIEW.GAME_START;
jest.mock("../../../../hooks/context", () => ({
  useToneAudioContext: jest.fn(() => {
    return { Tone: toneMock };
  }),
  useGameContext: jest.fn(() => {}),
  // useGameServerState: jest.fn(() => {
  //   return [mockGameStartView, {}, () => {}];
  // }),
  useSamplePlayer: jest.fn(() => [() => {}, () => {}, () => {}]),
}));

describe("renders Game component", () => {
  beforeEach(() => {
    mocked(useToneAudioContext).mockReturnValueOnce({ Tone: toneMock });
  });
  test("in game start view", () => {
    // mocked(useGameServerState).mockReturnValueOnce([
    //   mockGameStartView,
    //   {
    //     gameRules: gameRulesMock,
    //     gameView: "game_start",
    //   },
    //   pushMessageMock,
    // ]);

    const musicianId = "123";
    const gameComponent = render(
      <Game gameChannel={mockedChannel} musicianId={musicianId} />
    ).baseElement;

    expect(gameComponent).toMatchSnapshot();
  });

  // test("in round start view", () => {
  //   mocked(useToneAudioContext).mockReturnValueOnce({ Tone: toneMock });
  //   const musicianId = "123";
  //   const gameComponent = render(
  //     <Game gameChannel={mockedChannel} musicianId={musicianId} />
  //   ).baseElement;

  //   expect(gameComponent).toMatchSnapshot();
  // });
});
