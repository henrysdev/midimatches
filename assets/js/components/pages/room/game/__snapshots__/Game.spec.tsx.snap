// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Game component in game start view 1`] = `
<body>
  <div>
    <div>
      <h3>
        GameStart View
      </h3>
      <button>
        Ready Up
      </button>
    </div>
    <div>
      <div>
        Musician: 
        123
      </div>
      <pre />
    </div>
  </div>
</body>
`;
